src/brod_consumer.erl

```diff
@@ -48,265 +42,370 @@
+                          },
+      %% 4. Start consuming
+      {ok, CorrId} = send_fetch_request(State),
+      {noreply, State#state{corr_id = CorrId}}
   end;
-handle_call(stop, _From, State) ->
-  {stop, normal, ok, State};
-handle_call(get_socket, _From, State) ->
-  {reply, {ok, State#state.socket}, State};
-handle_call(Request, _From, State) ->
-  {stop, {unsupported_call, Request}, State}.
-
-handle_cast(Msg, State) ->
-  {stop, {unsupported_cast, Msg}, State}.
-
-handle_info({msg, _Pid, _CorrId, #fetch_response{} = R}, State0) ->
-  case handle_fetch_response(R, State0) of
-    {error, Error} ->
-      {stop, {broker_error, Error}, State0};
-    {ok, State} ->
-      MessageSet = brod_utils:fetch_response_to_message_set(R),
-      exec_callback(State#state.callback, MessageSet),
-      ok = send_fetch_request(State),
-      {noreply, State};
-    {empty, State} ->
-      erlang:send_after(State#state.sleep_timeout, self(), ?SEND_FETCH_REQUEST),
-      {noreply, State}
+handle_info({msg, _Pid, CorrId, R}, #state{corr_id = CorrId} = State0) ->
+  #fetch_response{topics = [TopicFetchData]} = R,
+  #topic_fetch_data{ topic = Topic
+                   , partitions = [PM]} = TopicFetchData,
+  #partition_messages{ partition = Partition
+                     , error_code = ErrorCode
+                     , high_wm_offset = HighWmOffset
+                     , last_offset = LastOffset
+                     , messages = Messages} = PM,
+  CbMod = State0#state.cb_mod,
+  CbState = State0#state.cb_state,
+  case brod_kafka:is_error(ErrorCode) of
+    true ->
+      Error = {ErrorCode, brod_kafka_errors:desc(ErrorCode)},
+      case CbMod:handle_error(Topic, Partition, Error, CbState) of
+        {ok, NewCbState} ->
+          erlang:send_after(?ERROR_COOLDOWN, self(), ?SEND_FETCH_REQUEST),
+          {noreply, State0#state{cb_state = NewCbState}};
+        {ok, NewCbState, NewCbOptions} ->
+          {ok, State} = update_cb_options(NewCbOptions, State0),
+          OldOffset = State0#state.offset,
+          NewOffset0 = State#state.offset,
+          SocketPid = State#state.socket_pid,
+          {ok, NewOffset} = maybe_get_valid_offset(SocketPid, OldOffset,
+                                                   NewOffset0, Topic, Partition),
+          erlang:send_after(?ERROR_COOLDOWN, self(), ?SEND_FETCH_REQUEST),
+          {noreply, State#state{cb_state = NewCbState, offset = NewOffset}};
+        stop ->
+          {stop, Error, State0};
+        Other ->
+          {stop, {callback_error, Other, Error}, State0}
+      end;
+    false ->
+      SleepTimeout = State0#state.sleep_timeout,
+      case Messages of
+        [] when SleepTimeout =:= 0 ->
+          {ok, NewCorrId} = send_fetch_request(State0),
+          {noreply, State0#state{corr_id = NewCorrId}};
+        [] when SleepTimeout > 0 ->
+          erlang:send_after(SleepTimeout, self(), ?SEND_FETCH_REQUEST),
+          {noreply, State0};
+        [_|_] ->
+          F = fun() ->
```

Comment:
we can: instead of put monitor reference in cb_pending, put a {reference(), fun((CbState) -> NewCbState))} to the tail of the list.
then spawn a pid to evaluate the only head of the list.